pkg/limits/stream_metadata.go

```diff
@@ -103,76 +108,143 @@ func (s *streamMetadata) Usage(tenant string, fn UsageFunc) {
 		}
 	}
 }
-func (s *streamMetadata) Store(tenant string, partitionID int32, streamHash, recTotalSize uint64, recordTime, bucketStart, bucketCutOff int64) {
+
+func (s *streamMetadata) StoreIf(tenant string, streams map[int32][]Stream, maxActiveStreams uint64, cutoff, bucketStart, bucketCutOff int64) (map[Reason][]uint64, uint64) {
 	i := s.getStripeIdx(tenant)
 
 	s.locks[i].Lock()
 	defer s.locks[i].Unlock()
 
-	// Initialize tenant map if it doesn't exist
 	if _, ok := s.stripes[i][tenant]; !ok {
-		s.stripes[i][tenant] = make(map[int32][]Stream)
+		s.stripes[i][tenant] = make(map[int32]map[uint64]Stream)
 	}
 
-	// Initialize partition map if it doesn't exist
-	if s.stripes[i][tenant][partitionID] == nil {
-		s.stripes[i][tenant][partitionID] = make([]Stream, 0)
-	}
+	var (
+		exceedLimits  = make(map[Reason][]uint64)
+		ingestedBytes uint64
+	)
+	for partitionID, streams := range streams {
+		if _, ok := s.stripes[i][tenant][partitionID]; !ok {
+			s.stripes[i][tenant][partitionID] = make(map[uint64]Stream)
+		}
+
+		var (
+			activeStreams = 0
+			newStreams    = 0
+		)
+
+		// Count as active streams all stream that are not expired.
+		for _, stored := range s.stripes[i][tenant][partitionID] {
```

Comment:
I have a feeling this will be a 🔥 path, we might need to optimize it later 👍

Reply:
I noticed this too because on average we have ~20-30k streams per partitions. Let's address this as a counter per partition to be stored along the metadata in a follow up PR. I think this will immediately solve our problems.

Reply:
How do you intend to keep the count up to date over time?

Reply:
Nevermind I was thinking out loudly about a `active []map[string]map[int32]uint64` close to our stripes that could be counted up in `StoreCond` and down in `Evict`, but I totally forgot that evict is periodic. Sorry turns out to be a bad idea.